forge-ai/src/main/java/forge/ai/ability/CharmAi.java

```diff
@@ -93,13 +94,22 @@ protected boolean checkApiLogic(Player ai, SpellAbility sa) {
     }
 
     private List<AbilitySub> chooseOptionsAi(List<AbilitySub> choices, final Player ai, boolean isTrigger, int num,
-            int min, boolean allowRepeat) {
+            int min, int pawprintLimit, boolean allowRepeat) {
         List<AbilitySub> chosenList = Lists.newArrayList();
         AiController aic = ((PlayerControllerAi) ai.getController()).getAi();
+        int pawprintAmount = 0;
         // First pass using standard canPlayAi() for good choices
         for (AbilitySub sub : choices) {
             sub.setActivatingPlayer(ai, true);
             if (AiPlayDecision.WillPlay == aic.canPlaySa(sub)) {
+                if (pawprintLimit > 0) {
+                    int curPawprintAmount = AbilityUtils.calculateAmount(sub.getHostCard(), sub.getParamOrDefault("Pawprint", "0"), sub);
+                    if (pawprintAmount + curPawprintAmount > pawprintLimit) {
+                        continue;
+                    } else {
+                        pawprintAmount += curPawprintAmount;
+                    }
+                }
```

Comment:
@Agetian because the later Choices are more expensive for Pawprint (and Spree?) I think it might be a good idea to flip the choices list, so AI tries to pay for the most expensive choices first

Reply:
Agreed :) Did it for Pawprint at the moment, should we always do it instead, in case it matters for more stuff? (can you also tell me what Spree card you're referring to, please?)

Reply:
looks like some extra test changes sneaked in though

Reply:
fixed the sneaky changes